app/services/attribute_asserter.rb

```diff
@@ -122,6 +139,13 @@ def authn_context
     authn_request.requested_ial_authn_context
   end
 
+  def x509_data
+    @x509_data ||= begin
+      x509_hash = user_session[:decrypted_x509]
+      X509::Attributes.new_from_json(x509_hash) if x509_hash
+    end
```

Comment:
memoizing on a potentially nil value is tricky, I think it's best to add a line like this in the beginning. On the other hand it looks like the nil case is just a hash get access that's "wasted" so it's not particularly expensive or slow

```suggestion
  def x509_data
    return @x509_data if defined?(@x509_data)
    @x509_data ||= begin
      x509_hash = user_session[:decrypted_x509]
      X509::Attributes.new_from_json(x509_hash) if x509_hash
    end
```